### PR TITLE
Add ZIO Streaming

### LINKED ISSUE #200: ZIO Streaming
ZIO now supports streaming via scalaz-zio-streams module, so it should be possible to implement  `streamBodyToPublisher`, `publisherToStreamBody` and `publisherToBytes` methods in `AsyncHttpClientZioBackend`, to allow response streaming with ZIO backend.

ZIO also provides a scalaz-zio-interop-reactivestreams module that would help to develop this feature. 

Example : 

```scala
import scalaz.zio._
import scalaz.zio.stream._
import scalaz.zio.interop.reactiveStreams._

val stream: Stream[Throwable, ByteBuffer] = ???
stream.toPublisher
```

Or 

```scala
val p: Publisher[ByteBuffer] = ???
p.toStream
```

For `publisherToBytes`, it may be somethig like `p.toStream.foldLeft(ByteBuffer.allocate(0))(concatByteBuffers)`

### PR DESCRIPTION
**Pull request by loicdescotte**
Fixes #200

### CODE REVIEW

**Comment by loicdescotte** [comment]
I'm not sure about the `bufferSize` setting, but the tests are working. 
I push a new commit (or a new PR) to change this setting.

**Comment by iravid** [comment]
cc @runtologist

File: async-http-client-backend/zio/src/main/scala/com/softwaremill/sttp/asynchttpclient/zio/AsyncHttpClientZioBackend.scala [hunk 1]
```diff
@@ -13,29 +15,43 @@ import org.asynchttpclient.{
...
-  override protected def streamBodyToPublisher(s: Nothing): Publisher[ByteBuf] =
-    s // nothing is everything
+  val runtime = new DefaultRuntime {}
 
-  override protected def publisherToStreamBody(p: Publisher[ByteBuffer]): Nothing =
-    throw new IllegalStateException("This backend does not support streaming")
+  private val bufferSize = 16
+
+  override protected def streamBodyToPublisher(s: Stream[Throwable, ByteBuffer]): Publisher[ByteBuf] =
+    runtime.unsafeRunSync(s.map(Unpooled.wrappedBuffer).toPublisher).toEither.right.get
```
Comment (runtologist) [inline async-http-client-backend/zio/src/main/scala/com/softwaremill/sttp/asynchttpclient/zio/AsyncHttpClientZioBackend.scala]: `runtime.unsafeRunSync(...).toEither.right.get` can be simplified to `runtime.unsafeRun(...)`

File: async-http-client-backend/zio/src/main/scala/com/softwaremill/sttp/asynchttpclient/zio/AsyncHttpClientZioBackend.scala [hunk 2]
```diff
@@ -13,29 +15,43 @@ import org.asynchttpclient.{
...
+import scalaz.zio.stream._
+import scalaz.zio.interop.reactiveStreams._
 
 class AsyncHttpClientZioBackend private (asyncHttpClient: AsyncHttpClient, closeClient: Boolean)
-    extends AsyncHttpClientBackend[IO[Throwable, ?], Nothing](asyncHttpClient, IOMonadAsyncError, closeClient) {
+    extends AsyncHttpClientBackend[Task, Stream[Throwable, ByteBuffer]](asyncHttpClient, IOMonadAsyncError, closeClient) {
 
-  override protected def streamBodyToPublisher(s: Nothing): Publisher[ByteBuf] =
-    s // nothing is everything
+  val runtime = new DefaultRuntime {}
```
Comment (runtologist) [inline async-http-client-backend/zio/src/main/scala/com/softwaremill/sttp/asynchttpclient/zio/AsyncHttpClientZioBackend.scala]: How do you feel about injecting the `runtime`? A user of this code will need a runtime anyway, so we do not need to start another one.

File: docs/backends/summary.rst [hunk 3]
```diff
@@ -22,7 +22,7 @@ Class                            Response wrapper             Supported stream t
 ``AkkaHttpBackend``              ``scala.concurrent.Future``  ``akka.stream.scaladsl.Source[ByteString, Any]``
 ``AsyncHttpClientFutureBackend`` ``scala.concurrent.Future``  n/a
 ``AsyncHttpClientScalazBackend`` ``scalaz.concurrent.Task``   n/a
-``AsyncHttpClientZioBackend``    ``scalaz.zio.IO``            n/a
+``AsyncHttpClientZioBackend``    ``scalaz.zio.IO``            Stream[Throwable, ByteBuffer]
```
Comment (adamw) [inline docs/backends/summary.rst]: I think I would consider creating a separate module with the streams dependency - just as fs2 isn't part of the cats module (though it's the de facto streaming library for cats). People might not want to pull in the additional dependency. So maybe a `AsyncHttpClientZioStreamsBackend`?

**Comment by loicdescotte** [comment]
Thanks @runtologist and @adamw for the review :)
I'll try to fix ASAP

**Comment by loicdescotte** [comment]
@adamw @runtologist I've integrated all the suggested changes in the PR :) 
WDYT?

**Comment by runtologist** [comment]
LGTM. :)

**Comment by adamw** [comment]
Released in 1.5.16. Thanks! :)

**Comment by loicdescotte** [comment]
@Voltir @adamw oops sorry for this, thanks for the report and the fix :)

**Pull request merged**